packages/php-wasm/universal/src/lib/php-request-handler.ts

```diff
@@ -415,14 +419,46 @@ export class PHPRequestHandler implements AsyncDisposable {
 				const possibleIndexPath = joinPaths(fsPath, possibleIndexFile);
 				if (primaryPhp.isFile(possibleIndexPath)) {
 					fsPath = possibleIndexPath;
+
+					// Include the resolved index file in the final rewritten request URL.
+					rewrittenRequestUrl.pathname = joinPaths(
+						rewrittenRequestUrl.pathname,
+						possibleIndexFile
+					);
+					break;
+				}
+			}
+		}
+
+		if (!primaryPhp.isFile(fsPath)) {
+			/**
+			 * Try resolving a partial path.
+			 *
+			 * Example:
+			 *
+			 * – Request URL: /file.php/index.php
+			 * – Document Root: /var/www
+			 *
+			 * If /var/www/file.php/index.php does not exist, but /var/www/file.php does,
+			 * use /var/www/file.php. This is also what Apache and PHP Dev Server do.
+			 */
+			let pathToTry = rewrittenRequestUrl.pathname;
+			while (true) {
+				pathToTry = dirname(pathToTry);
+				if (pathToTry === '/' || !pathToTry.includes('/')) {
+					// We've tried all segments for a partial path.
+					break;
+				}
+				if (primaryPhp.isFile(joinPaths(this.#DOCROOT, pathToTry))) {
+					fsPath = joinPaths(this.#DOCROOT, pathToTry);
 					break;
 				}
 			}
```

Comment:
Every time I see a `while (true)` loop that isn't a main program loop, I wonder whether it can be rewritten with an explicit loop condition and remain readable. Here's my attempt:

```suggestion
			for (
				let pathToTry = dirname(rewrittenRequestUrl.pathname);
				pathToTry.startsWith('/') && pathToTry !== dirname(pathToTry);
				pathToTry = dirname(pathToTry)
			) {
				if (primaryPhp.isFile(joinPaths(this.#DOCROOT, pathToTry))) {
					fsPath = joinPaths(this.#DOCROOT, pathToTry);
					break;
				}
			}
```

Reply:
Something in me protested when I saw the for() syntax used for something other than a counter, but I think you have a point and I rewrote this loop to use an explicit condition. Thank you!

```ts
			let pathToTry = dirname(rewrittenRequestUrl.pathname);
			while (
				pathToTry.startsWith('/') &&
				pathToTry !== dirname(pathToTry)
			) {
				pathToTry = dirname(pathToTry);
				if (primaryPhp.isFile(joinPaths(this.#DOCROOT, pathToTry))) {
					fsPath = joinPaths(this.#DOCROOT, pathToTry);
					break;
				}
			}
```

Reply:
> Something in me protested when I saw the for() syntax used for something other than a counter, but I think you have a point and I rewrote this loop to use an explicit condition. Thank you!

😄 I like how this bounced between our sensibilities to a good result. Thanks